Sources/NIOHTTP1Client/main.swift

```diff
@@ -0,0 +1,89 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+import NIOOpenSSL
+
+private final class HTTPResponseHandler: ChannelInboundHandler {
+
+    let promise: EventLoopPromise<Void>
+
+    init(_ promise: EventLoopPromise<Void>) {
+        self.promise = promise
+    }
+
+    typealias InboundIn = HTTPClientResponsePart
+
+    func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+        let httpResponsePart = unwrapInboundIn(data)
+        switch httpResponsePart {
+        case .head(let httpResponseHeader):
+            print("\(httpResponseHeader.version) \(httpResponseHeader.status.code) \(httpResponseHeader.status.reasonPhrase)")
+            for (name, value) in httpResponseHeader.headers {
+                print("\(name): \(value)")
+            }
+        case .body(var byteBuffer):
+            if let responseBody = byteBuffer.readString(length: byteBuffer.readableBytes) {
+                print(responseBody)
+            }
+        case .end(_):
+            break
+        }
+    }
+
+    func channelReadComplete(ctx: ChannelHandlerContext) {
+        _ = ctx.channel.close()
+        promise.succeed(result: ())
+    }
+
+    func errorCaught(ctx: ChannelHandlerContext, error: Error) {
+        print("Error: ", error)
+        _ = ctx.channel.close()
+        promise.succeed(result: ())
+    }
+}
+
+let eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+let promise: EventLoopPromise<Void> = eventLoopGroup.next().newPromise()
+defer {
+    try! promise.futureResult.wait()
+    try! eventLoopGroup.syncShutdownGracefully()
+}
+
+let tlsConfiguration = TLSConfiguration.forClient()
+let sslContext = try! SSLContext(configuration: tlsConfiguration)
+let openSslHandler = try! OpenSSLClientHandler(context: sslContext, serverHostname: "httpbin.org")
+
+let bootstrap = ClientBootstrap(group: eventLoopGroup)
+        .channelOption(ChannelOptions.socket(SocketOptionLevel(SOL_SOCKET), SO_REUSEADDR), value: 1)
+        .channelInitializer { channel in
+            _ = channel.pipeline.add(handler: openSslHandler)
```

Comment:
1. the `openSslHandler` should be constructed inside of the `channelInitializer` closure as it might be run more than once (see the [warning in here](https://apple.github.io/swift-nio/docs/current/NIO/Classes/ClientBootstrap.html#/s:3NIO15ClientBootstrapC18channelInitializerACXDAA15EventLoopFutureCyytGAA7Channel_pcF) )
2. the futures should be properly chained:

```swift
[...]
.channelInitializer { channel in
    return channel.pipeline.add(handler: OpenSSLClientHandler(...)).then {
        channel.pipeline.addHTTPClientHandlers()
    }.then {
        channel.pipeline.add(handler: HTTPResponseHandler(promise))
    }
}
```

Reply:
Thanks! I fixed them. Please see 6e21fc7c597f7dd6aabf1669f28b07ab579d2121 .